milestones/api.py

```diff
@@ -0,0 +1,312 @@
+        exceptions.raise_exception(
+            "MilestoneRelationshipType",
+            name,
+            exceptions.InvalidMilestoneRelationshipTypeException
+        )
+
+
+def _validate_user(user):
+    """ Validation helper """
+    if not validators.user_is_valid(user):
+        exceptions.raise_exception(
+            "User",
+            user,
+            exceptions.InvalidUserException
+        )
+
+
+# PUBLIC FUNCTIONS
+def add_milestone(milestone):
+    """
+    Passes a new milestone to the data layer for storage
+    """
+    _validate_milestone(milestone)
+    milestone = data.create_milestone(milestone)
+    return milestone
+
+
+def edit_milestone(milestone):
+    """
+    Passes an updated milestone to the data layer for storage
+    """
+    _validate_milestone(milestone)
+    try:
+        return data.update_milestone(milestone)
+    except exceptions.InvalidMilestoneException:
+        raise
+
+
+def get_milestone(milestone_id):
+    """
+    Retrieves the specified milestone
+    """
+    milestone = {
+        'id': milestone_id,
+    }
+    milestones = data.fetch_milestones(milestone)
+    if not len(milestones):
+        return None
+    return milestones[0]
+
+
+def get_milestones(namespace):
+    """
+    Retrieves the specified milestone by namespace
+    (no other way to really do it right now)
+    """
+    milestone = {
+        'namespace': namespace,
+    }
+    _validate_milestone(milestone)
+    return data.fetch_milestones(milestone)
+
+
+def remove_milestone(milestone_id):
+    """
+    Removes the specified milestone
+    """
+    milestone = {
+        'id': milestone_id,
+    }
+    data.delete_milestone(milestone)
+
+
+def add_course_milestone(course_key, relationship, milestone):
+    """
+    Adds a course-milestone link to the system
+    'relationship': string value (eg: 'requires')
+    """
+    _validate_course_key(course_key)
+    _validate_milestone_relationship_type(relationship)
+    _validate_milestone(milestone)
+    data.create_course_milestone(course_key=course_key, relationship=relationship, milestone=milestone)
+
+
+def get_course_milestones(course_key, relationship=None):
+    """
+    Retrieves the set of milestones for a given course
+    'relationship': optional filter on milestone relationship type (string, eg: 'requires')
+    Returns an array of dicts containing milestones
+    """
+    _validate_course_key(course_key)
+
+    if relationship is not None:
+        _validate_milestone_relationship_type(relationship)
+    return data.fetch_courses_milestones(course_keys=[course_key], relationship=relationship)
+
+
+def get_course_required_milestones(course_key, user):
+    """
+    Retrieves the set of required milestones for a given course that a user has not yet collected
+    """
+    _validate_course_key(course_key)
+    _validate_user(user)
+    required_milestones = data.fetch_courses_milestones([course_key], 'requires', user)
```

Comment:
rather than a in code string, should we have a defined constant for 'requires'. It seems like that is a system defined relationship. I see this in a few places.